airbyte-integrations/bases/base-typing-deduping/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/AirbyteTypeUtils.java

```diff
@@ -0,0 +1,157 @@
+      AirbyteProtocolType.BOOLEAN,
+      AirbyteProtocolType.INTEGER,
+      AirbyteProtocolType.NUMBER,
+      AirbyteProtocolType.TIMESTAMP_WITHOUT_TIMEZONE,
+      AirbyteProtocolType.TIMESTAMP_WITH_TIMEZONE,
+      AirbyteProtocolType.DATE,
+      AirbyteProtocolType.TIME_WITH_TIMEZONE,
+      AirbyteProtocolType.TIME_WITHOUT_TIMEZONE,
+      AirbyteProtocolType.STRING);
+
+  protected static boolean nodeIsType(final JsonNode node, final String type) {
+    if (node == null || !node.isTextual()) {
+      return false;
+    }
+    return node.equals(TextNode.valueOf(type));
+  }
+
+  private static boolean nodeIsOrContainsType(final JsonNode node, final String type) {
+    if (node == null) {
+      return false;
+    } else if (node.isTextual()) {
+      return nodeIsType(node, type);
+    } else if (node.isArray()) {
+      for (final JsonNode element : node) {
+        if (nodeIsType(element, type)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  protected static AirbyteType getAirbyteProtocolType(final JsonNode node) {
+    final JsonNode propertyType = node.get("type");
+    final JsonNode airbyteType = node.get("airbyte_type");
+    final JsonNode format = node.get("format");
+    if (Stream.of(propertyType, airbyteType, format).allMatch(Objects::isNull)) {
+      return AirbyteProtocolType.matches(node.asText());
+    }
+
+    if (nodeIsOrContainsType(propertyType, "boolean")) {
+      return AirbyteProtocolType.BOOLEAN;
+    } else if (nodeIsOrContainsType(propertyType, "integer")) {
+      return AirbyteProtocolType.INTEGER;
+    } else if (nodeIsOrContainsType(propertyType, "number")) {
+      if (nodeIsType(airbyteType, "integer")) {
+        return AirbyteProtocolType.INTEGER;
+      } else {
+        return AirbyteProtocolType.NUMBER;
+      }
+    } else if (nodeIsOrContainsType(propertyType, "string")) {
+      if (nodeIsOrContainsType(format, "date")) {
+        return AirbyteProtocolType.DATE;
+      } else if (nodeIsType(format, "time")) {
+        if (nodeIsType(airbyteType, "timestamp_without_timezone")) {
+          return AirbyteProtocolType.TIME_WITHOUT_TIMEZONE;
+        } else if (nodeIsType(airbyteType, "timestamp_with_timezone")) {
+          return AirbyteProtocolType.TIME_WITH_TIMEZONE;
+        }
+      } else if (nodeIsOrContainsType(format, "date-time")) {
+        if (nodeIsType(airbyteType, "timestamp_without_timezone")) {
+          return AirbyteProtocolType.TIMESTAMP_WITHOUT_TIMEZONE;
+        } else if (airbyteType == null || nodeIsType(airbyteType, "timestamp_with_timezone")) {
+          return AirbyteProtocolType.TIMESTAMP_WITH_TIMEZONE;
+        }
+      } else {
+        return AirbyteProtocolType.STRING;
+      }
+    }
+
+    return AirbyteProtocolType.UNKNOWN;
+  }
+
+  // Pick which type in a OneOf has precedence
+  public static AirbyteType chooseOneOfType(final OneOf o) {
```

Comment:
I find this method confusing. From the naming of the method it starts off doing what I would expect, we have a `OneOf` object which has an `options` field which contains several airbyte types, presumable one of these is what the return value of this method should be. It looks like then there is some special handling where if there is an array or struct, we should select those options regardless of other options (in the order of array first, then struct). Finally, we loop through the protocolTypes in order, check whether that was found in the options but compare that against a set of preferred types (the excluded types) eventually returning the appropriate type?

Here is an (untested) suggested rewrite
```agsl
final List<AirbyteType> options = o.options();

// handle arrays and structs immediately

Optional<AirbyteType> potentialArray = options.stream().filter(o -> o instanceof  Array).findAny();
if (potentialArray.isPresent()) {
  return potentialArray.get()
}
Optional<AirbyteType> potentialStruct = options.stream().filter(o -> o instanceof  Struct).findAny();
if (potentialStruct.isPresent()) {
  return potentialStruct.get()
}

// at this point, we only care about airbyte protocol types, sort the options in our order of preference
// reading the rest of this class, it seems like the order of preference is "most broad" i.e. we choose STRING over BOOLEAN, 
// this comparator isn't great, but maybe its close? (the reverse is because the current ordering is "reversed" to this comparator)

return options.stream()
  .filter(o -> instanceof AirbyteProtocolType)
  .map(o -> (AirbyteProtocolType) o)
  .sorted(Comparator.comparing(t -> ORDERED_PROTOCOL_TYPES.indexOf(t), Comparator.reverseOrder()))
  .findFirst()
  .orElse(AirbyteProtocolType.UNKNOWN)


```

Reply:
@jbfbell Thanks for the feedback and code suggestion! Trying to figure out how to incorporate the special type exclusion logic into the stream, and also how this affects the efficiency discussion 🤔